lib/os/mempool.c

```diff
@@ -179,10 +167,8 @@ static unsigned int bfree_recombine(struct sys_mem_pool_base *p, int level,
 		for (i = 0; i < 4; i++) {
 			int b = (bn & ~3) + i;
 
-			if (block_fits(p, block_ptr(p, lsz, b), lsz)) {
-				clear_free_bit(p, level, b);
-				sys_dlist_remove(block_ptr(p, lsz, b));
-			}
+			clear_free_bit(p, level, b);
+			sys_dlist_remove(block_ptr(p, lsz, b));
```

Comment:
Not understanding this bit here.  The reason for the fits check is that you might be splitting an arbitrary-sized superblock into 4 children, and find that because of alignment only three aligned subblocks are going to fit.  Superblocks are regions specified by the user, they aren't power-of-four sized in the general case.

That is, consider specifying a 248 block mempool area you want to use for a heap and allocating an 8 byte region.  The first step is to split that superblock into four children at offsets 0, 64, 128, and 192.  But that last block doesn't fit!  The other three are legal though.